[Site Isolation] Mark scrolling tree nodes that are parents of a RemoteFrame

nmoucht opened the pull request:
#### 385c72e3a6462abfbbdbee6b77a344a663ffd74d
<pre>
[Site Isolation] Mark scrolling tree nodes that are parents of a RemoteFrame
<a href="https://bugs.webkit.org/show_bug.cgi?id=268518">https://bugs.webkit.org/show_bug.cgi?id=268518</a>
<a href="https://rdar.apple.com/122054689">rdar://122054689</a>

Reviewed by Simon Fraser.

This is part 1 of splitting up <a href="https://github.com/WebKit/WebKit/pull/23242.">https://github.com/WebKit/WebKit/pull/23242.</a> The
purpose of this patch is to mark scrolling tree nodes that have an associated
RemoteFrame, currently using the LayerHostingContextIdentifier. This will later
be used to connect the scrolling tree from the iframe process in the ui process.

* Source/WebCore/page/FrameView.h:
* Source/WebCore/page/RemoteFrame.cpp:
(WebCore::RemoteFrame::didFinishLoadInAnotherProcess):
* Source/WebCore/page/RemoteFrame.h:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::setLayerHostingContextIdentifier):
(WebCore::AsyncScrollingCoordinator::frameViewForScrollingNode const):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
(WebCore::ScrollingCoordinator::isSiteIsolatedTree):
(WebCore::ScrollingCoordinator::setLayerHostingContextIdentifier):
* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.cpp:
(WebCore::ScrollingStateFrameScrollingNode::ScrollingStateFrameScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStateNode.cpp:
(WebCore::operator&lt;&lt;):
(WebCore::ScrollingStateNode::setLayer):
* Source/WebCore/page/scrolling/ScrollingStateNode.h:
(WebCore::LayerRepresentation::operator PlatformLayer* const):
(WebCore::LayerRepresentation::toRepresentation const):
(WebCore::LayerRepresentation::representation const):
* Source/WebCore/page/scrolling/ScrollingStateOverflowScrollingNode.cpp:
(WebCore::ScrollingStateOverflowScrollingNode::ScrollingStateOverflowScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStateOverflowScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.cpp:
(WebCore::ScrollingStatePluginScrollingNode::ScrollingStatePluginScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
(WebCore::ScrollingStateScrollingNode::ScrollingStateScrollingNode):
(WebCore::ScrollingStateScrollingNode::setLayerHostingContextIdentifier):
(WebCore::ScrollingStateScrollingNode::dumpProperties const):
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
(WebCore::ScrollingStateScrollingNode::layerHostingContextIdentifier const):
* Source/WebCore/page/scrolling/ScrollingStateTree.cpp:
(WebCore::ScrollingStateTree::createUnparentedNode):
(WebCore::ScrollingStateTree::insertNode):
* Source/WebCore/page/scrolling/ScrollingStateTree.h:
* Source/WebCore/page/scrolling/ScrollingTree.h:
* Source/WebCore/page/scrolling/ScrollingTreeNode.cpp:
(WebCore::ScrollingTreeNode::removeAllChildren):
* Source/WebCore/page/scrolling/ScrollingTreeNode.h:
(WebCore::ScrollingTreeNode::setIsHostedSubtree):
(WebCore::ScrollingTreeNode::isHostedSubtree const):
(WebCore::ScrollingTreeNode::layerHostingContextIdentifier):
(WebCore::ScrollingTreeNode::setlayerHostingContextIdentifier):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
(WebCore::ScrollingTreeScrollingNode::commitStateBeforeChildren):
(WebCore::ScrollingTreeScrollingNode::dumpProperties const):
* Source/WebCore/page/scrolling/mac/ScrollingTreeMac.mm:
(scrollingNodeIDForLayer):
(ScrollingTreeMac::scrollingNodeForPoint):

Canonical link: <a href="https://commits.webkit.org/274123@main">https://commits.webkit.org/274123@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/be84f8ccad4115e7151406f994c7a6f02c880285

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37936 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16844 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40194 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40480 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33754 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/39294 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19540 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14146 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32077 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38509 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14192 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33227 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12389 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12347 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/33957 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41753 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34422 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34424 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38210 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12928 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10569 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36393 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14502 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8518 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13359 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/13954 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/RemoteFrame.cpp:
```diff
@@ -81,8 +84,18 @@ void RemoteFrame::didFinishLoadInAnotherProcess()
+            return;
+
+        if (auto frameHostingNodeID = widgetRenderer->layer()->backing()->scrollingNodeIDForRole(ScrollCoordinationRole::FrameHosting))
+            page()->scrollingCoordinator()->setLayerHostingContextIdentifier(frameHostingNodeID, m_layerHostingContextIdentifier);
```
This probably isn't the best place to set this. What if this iframe is unparented and re-parented (needs a test)? Do we need to hook things up before the load finishes (pages can display content before the load event fires).

I think it would make more sense to call this in `RenderLayerCompositor::updateScrollingNodeForFrameHostingRole`.

nmoucht replied:
Right fixed.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
```diff
@@ -75,6 +75,8 @@ class AsyncScrollingCoordinator : public ScrollingCoordinator {
     
     WEBCORE_EXPORT void setMouseIsOverContentArea(ScrollableArea&, bool) override;
     WEBCORE_EXPORT void setMouseMovedInContentArea(ScrollableArea&) override;
+    WEBCORE_EXPORT void setLayerHostingContextIdentifier(ScrollingNodeID, Markable<LayerHostingContextIdentifier>) override;
```
Let's call this `setLayerHostingContextIdentifierForFrameHostingNode()`. It's better to use `std::optional<?>` for arguments.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -421,6 +421,14 @@ void AsyncScrollingCoordinator::setMouseMovedInContentArea(ScrollableArea& scrol
+void AsyncScrollingCoordinator::setLayerHostingContextIdentifier(ScrollingNodeID scrollingNodeID, Markable<LayerHostingContextIdentifier> identifier)
+{
+    auto stateNode = dynamicDowncast<ScrollingStateFrameHostingNode>(m_scrollingStateTree->stateNodeForID(scrollingNodeID));
+    if (!stateNode)
```
This should assert, since it's an error to call this on something that isn't a frame hosting node.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingTree.h:
```diff
@@ -28,6 +28,7 @@
 #if ENABLE(ASYNC_SCROLLING)
 
 #include "EventTrackingRegions.h"
+#include "LayerHostingContextIdentifier.h"
```
Can you forward-declare instead?

nmoucht replied:
Removed.

smfr reviewed (approved):
Can we write a test for this?

nmoucht committed:
[Site Isolation] Mark scrolling tree nodes that are parents of a RemoteFrame
https://bugs.webkit.org/show_bug.cgi?id=268518
rdar://122054689

Reviewed by Simon Fraser.

This is part 1 of splitting up https://github.com/WebKit/WebKit/pull/23242. The
purpose of this patch is to mark scrolling tree nodes that have an associated
RemoteFrame, currently using the LayerHostingContextIdentifier. This will later
be used to connect the scrolling tree from the iframe process in the ui process.

* Source/WebCore/page/FrameView.h:
* Source/WebCore/page/RemoteFrame.cpp:
(WebCore::RemoteFrame::didFinishLoadInAnotherProcess):
* Source/WebCore/page/RemoteFrame.h:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::setLayerHostingContextIdentifier):
(WebCore::AsyncScrollingCoordinator::frameViewForScrollingNode const):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
(WebCore::ScrollingCoordinator::isSiteIsolatedTree):
(WebCore::ScrollingCoordinator::setLayerHostingContextIdentifier):
* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.cpp:
(WebCore::ScrollingStateFrameScrollingNode::ScrollingStateFrameScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStateNode.cpp:
(WebCore::operator<<):
(WebCore::ScrollingStateNode::setLayer):
* Source/WebCore/page/scrolling/ScrollingStateNode.h:
(WebCore::LayerRepresentation::operator PlatformLayer* const):
(WebCore::LayerRepresentation::toRepresentation const):
(WebCore::LayerRepresentation::representation const):
* Source/WebCore/page/scrolling/ScrollingStateOverflowScrollingNode.cpp:
(WebCore::ScrollingStateOverflowScrollingNode::ScrollingStateOverflowScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStateOverflowScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.cpp:
(WebCore::ScrollingStatePluginScrollingNode::ScrollingStatePluginScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
(WebCore::ScrollingStateScrollingNode::ScrollingStateScrollingNode):
(WebCore::ScrollingStateScrollingNode::setLayerHostingContextIdentifier):
(WebCore::ScrollingStateScrollingNode::dumpProperties const):
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
(WebCore::ScrollingStateScrollingNode::layerHostingContextIdentifier const):
* Source/WebCore/page/scrolling/ScrollingStateTree.cpp:
(WebCore::ScrollingStateTree::createUnparentedNode):
(WebCore::ScrollingStateTree::insertNode):
* Source/WebCore/page/scrolling/ScrollingStateTree.h:
* Source/WebCore/page/scrolling/ScrollingTree.h:
* Source/WebCore/page/scrolling/ScrollingTreeNode.cpp:
(WebCore::ScrollingTreeNode::removeAllChildren):
* Source/WebCore/page/scrolling/ScrollingTreeNode.h:
(WebCore::ScrollingTreeNode::setIsHostedSubtree):
(WebCore::ScrollingTreeNode::isHostedSubtree const):
(WebCore::ScrollingTreeNode::layerHostingContextIdentifier):
(WebCore::ScrollingTreeNode::setlayerHostingContextIdentifier):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
(WebCore::ScrollingTreeScrollingNode::commitStateBeforeChildren):
(WebCore::ScrollingTreeScrollingNode::dumpProperties const):
* Source/WebCore/page/scrolling/mac/ScrollingTreeMac.mm:
(scrollingNodeIDForLayer):
(ScrollingTreeMac::scrollingNodeForPoint):

Canonical link: https://commits.webkit.org/274123@main
